### PR TITLE
1.0.0

### LINKED ISSUE #115: Edit/delete messages
Each message should have a way to open a dropdown (probably an ellipsis-button to the side) letting the user edit or delete it.

* Edit (if the user owns the message). See #58 for old discussion on how editing a message should work. @PullJosh? On Gitter and Discord, editing a message replaces the message's content with an input box for typing in; that seems the most intuitive to me. (Also, it doesn't remove existing content from the chat input - it creates a *new* input.)

* Delete (if the user owns the message, or is an admin). Should there be a confirmation dialog?

The actual dropdown should be implemented as a "separate" unit - other functions should be able to easily use the same dropdown code. I'm thinking there should be a global dropdown element which is filled/shown by a `showDropdown` function, that way we don't need to create and keep track of tooooooooooons of DOM elements (a dropdown and its items for every single message sent!).

### LINKED ISSUE #234: `.Link`s aren't styled on Chrome
No idea why!

### LINKED ISSUE #261: "Disconnected" indicator

### LINKED ISSUE #276: User list should sort without considering letter case
Right now, `PullJosh` sorts before `eq`!

### LINKED ISSUE #281: Implement API spec 1.0.0
[The spec](https://github.com/decent-chat/decent/blob/1.0.0/docs/api.md).

### LINKED ISSUE #305: Stop using Libravatar
Libravatar is shutting down on 2018-09-01[(blog post)](https://blog.libravatar.org/posts/Libravatar.org_is_shutting_down_on_2018-09-01/). We'll need to stop using it soon, and prepare to switch to something else as the main source of Avatars. I'm guessing it'll be Gravatar.

### PR DESCRIPTION
**Pull request by bates64**
Ready to merge as soon as https://github.com/decent-chat/decent/projects/7 is fully complete :)

Resolves #115
Fixes #234
Fixes #261
Fixes #276
Fixes #281
Fixes #305

### CODE REVIEW

File: packages/client/js/storage.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  else storage.setItem(key, JSON.stringify(value))
+
+  console.log(`Storage: set ${key} =`, value)
+}
+
+function load(key, defaultValue) {
+  const str = storage.getItem(key)
+  const value = str ? JSON.parse(str) : defaultValue
+
+  console.log(`Storage: loaded ${key} =`, value)
```
Comment (towerofnix) [inline packages/client/js/storage.js]: Don't keep these in the final 1.0.0!

File: packages/server/api.js [hunk 2]
```diff
@@ -742,6 +742,30 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
...
+  /*
+  app.patch('/api/channels/:channelID/role-permissions', [
+    ...middleware.loadVarFromParams('channelID'),
+    ...middleware.getChannelFromID('channelID', 'channel'),
+
+    (request, response) => {
+      const { channel } = request[middleware.vars]
+    }
+  ])
+  */
```
Comment (heyitsmeuralex) [inline packages/server/api.js]: um

Comment (towerofnix) [reply packages/server/api.js]: WIP, of course.

File: packages/server/test/api-messages.js [hunk 3]
```diff
@@ -47,6 +47,21 @@ test('POST /api/messages', t => {
...
+          channelID, sessionID: senderSID,
+          text: 'Important message important message important message aaaaa',
+          type: 'system'
+        })
+      })
+      t.fail('Could post system message without sendSystemMessages permission')
+    } catch (error) {
+      t.is(error.code, 'NOT_ALLOWED')
+      t.is(error.missingPermission, 'sendSystemMessages')
+    }
```
Comment (towerofnix) [inline packages/server/test/api-messages.js]: @heyitsmeuralex You still need to update `serialize.message`, and you should update `test/serialize.js` too.

**Review by towerofnix — COMMENTED** [review]
Just a nit and a question re: flair test. System messages look good otherwise!

File: packages/server/serialize.js [hunk 4]
```diff
@@ -7,9 +7,9 @@ module.exports = function makeSerializers({util, db}) {
   const serialize = {
     message: async m => ({
       id: m._id,
-      authorUsername: m.authorUsername,
-      authorID: m.authorID,
-      authorAvatarURL: emailToAvatarURL(m.authorEmail || m.authorID),
+      authorUsername: m.type === 'system' ? null : m.authorUsername,
+      authorID: m.type === 'system' ? null : m.authorID,
+      authorAvatarURL: m.type === 'system' ? null : emailToAvatarURL(m.authorEmail || m.authorID),
```
[the remark this answers is not in the document]
Comment (heyitsmeuralex) [reply packages/server/serialize.js]: If we ever add more types they'll likely be tied to users too - `system` is the special case, not `user`.

**Pull request closed without merging**